ibis/expr/operations/relations.py

```diff
@@ -606,14 +606,27 @@ def _dedup_join_columns(expr: ir.Table, lname: str, rname: str):
     overlap = frozenset(column for column in left.columns if column in right_columns)
     equal = set()
 
-    if isinstance(op, InnerJoin) and util.all_of(op.predicates, Equals):
-        # For inner joins composed exclusively of equality predicates, we can
-        # avoid renaming columns with colliding names if their values are
-        # guaranteed to be equal due to the predicate. Here we collect a set of
-        # colliding column names that are known to have equal values between
-        # the left and right tables in the join.
+    # For inner joins composed exclusively of equality predicates, we can
+    # avoid renaming columns with colliding names if their values are
+    # guaranteed to be equal due to the predicate. Here we collect a set of
+    # colliding column names that are known to have equal values between
+    # the left and right tables in the join.
+    #
+    # TODO (mehmet): Included also asof joins, check with the reviewers about
+    # the implications!
+    # Edit: This change leads to assert failures in tests for other backends.
+    # Q. Why was the deduplication of columns implemented only for `InnerJoin`?
```

Comment:
We need clarification on this question:
In `_dedup_join_columns()`, why was the deduplication of columns implemented only for `InnerJoin`?

Reply:
The only case where the key columns can be deduplicated without loss of information is when:

1. The join is an inner join
2. The predicates are all equality predicates

If either of those things is false then deduplicating columns loses information since the join keys are not equivalent.